src/chatbot.py

```diff
@@ -129,17 +128,42 @@ def build_model(checkpoint_name):
 
 
 
-def train(pairs, encoder, decoder, embedding, personas, word_map, person_map, checkpoint):
-    # Ensure dropout layers are in train mode
-    encoder.train()
-    decoder.train()
-
+def train(mode, encoder, decoder, embedding, personas, word_map, person_map, checkpoint):
     trainer = Trainer(encoder, decoder, word_map, person_map, embedding, personas)
 
     if checkpoint:
         trainer.load(checkpoint)
 
-    trainer.train(pairs, config.N_ITER, config.BATCH_SIZE)
+    if mode == 'pretrain':
+        corpus = config.PRETRAIN_CORPUS
+        trim_corpus = True
+        train_persona = False
+        n_iter = config.PRETRAIN_N_ITER
+
+    elif mode == 'finetune':
+        # finetune requires checkpoint
+        assert checkpoint is not None
+
+        corpus = config.FINETUNE_CORPUS
+        trim_corpus = False
+        train_persona = True
+        n_iter = config.FINETUNE_N_ITER
+
+        # finetune from pretrain checkpoint, reset start iter
+        # verify stage keyword to make it compatible with previous
+        if 'stage' not in checkpoint or checkpoint['stage'] != 'finetune':
+            trainer.reset_iter()
```

Comment:
@jiahao42 here, say we pretrained for 100K iteration, now I only want to `finetune` for another `10K` iteration with `pretrain` checkpoint, so i need to reset the `100K` iter to `1`

but if I `finetune` with a `finetune` checkpoint, just continue training. same for `pretrain` with `pretrain` checkpoint

Reply:
of course, one should never `pretrain` with `finetune` checkpoint......

Reply:
Oh sorry, I didn't see you will reset iter in fine tune here.